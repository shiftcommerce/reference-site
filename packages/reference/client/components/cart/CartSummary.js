// Libraries
import { Component } from 'react'
import Link from 'next/link'

// Libs
import { calculateCartSummary } from '../../lib/calculateCartSummary'
import { fixedPrice } from '../../lib/fixedPrice'

// Objects
import Button from '../../objects/Button'

class CartSummary extends Component {
  shippingText () {
    if (typeof window === 'undefined' || window.location.pathname !== '/checkout') {
      // TODO: make this dynamic in future
      return <>
        <a>Estimated shipping cost: &pound;3.45</a>
      </>
    } else {
      return <>
         <a>Shipping cost:</a>
       </>
    }
  }

  render () {
    const totals = calculateCartSummary(this.props.cart)

    return (
      <>
      <section className='c-cart-summary'>
        <dl>
          <dt><a>Total Price:</a></dt>
          <dd><a>&pound;{fixedPrice(totals.subTotal)}</a></dd>
        </dl>
        <dl>
          <dt>{ this.shippingText() }</dt>
        </dl>
        <dl>
          <dt><a className='u-bold'>TOTAL:</a></dt>
          <dd><a className='u-bold'>&pound;{fixedPrice(totals.total)}</a></dd>
        </dl>
        <dl>
          <a>* Including VAT</a>
        </dl>
      </section>
      <section className='c-cart-summary__buttons'>
        <Button
          aria-label='Continue Shopping'
          label='Continue shopping'
          size='lrg'
          status='grey'
          className='c-cart-summary__buttons--continue'
          type='button'
        />
        <Link href='/checkout' as=''>
          <Button
            aria-label='Place Order'
            label='Continue to payment'
            size='lrg'
            className='c-cart-summary__buttons--proceed'
            status='positive'
            type='button'
          />
        </Link>
      </section>
      </>
    )
  }
}

export default CartSummary
