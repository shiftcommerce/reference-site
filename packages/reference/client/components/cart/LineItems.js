// Libraries
import { Component } from 'react'
import Link from 'next/link'
import { fixedPrice } from '../../lib/fixedPrice'

// Objects
import Image from '../../objects/Image'

class LineItems extends Component {
  // TODO: extract this out into it's own service class
  renderOptions (lineItem) {
    let availableQuantity = Array.from(Array(Math.min(lineItem.stockAvailableLevel, 11)).keys())
    availableQuantity.shift()

    return <select value={lineItem.quantity} onChange={this.props.updateQuantity} data-variant={lineItem.sku}>
      {
        availableQuantity.map(index => {
          return <option key={index} value={index} aria-setsize={index} aria-posinset={index + 1}>
            {index}
          </option>
        })
      }
    </select>
  }

  renderButtonsAndTotal (lineItem) {
    return <>
      <div className='o-grid-item5__lineitems'>
        <div className='c-line-items__buttons'>
          <div className='c-line-items__anchor'>
            <a className='c-line-items__delete' data-variant={lineItem.sku} onClick={this.props.deleteItem} >
              Delete
            </a>
          </div>
        </div>
        <div className='c-line-items__total-wrapper'>
          <a className='c-line-items__total'>&pound;{fixedPrice(lineItem.price * lineItem.quantity)}</a>
        </div>
      </div>
    </>
  }

  renderTitle (lineItem) {
    return <>
      <div className='o-grid-item2__lineitems'>
        <div className='c-line-items__details'>
          <h4 className='c-line-items__details-title  u-bold'>
            {lineItem.title}
          </h4>
          <div className='c-line-items__details-sku'>
            <span>
              {lineItem.sku}
            </span>
          </div>
        </div>
      </div>
    </>
  }

  renderParams (lineItem) {
    return <>
      <div className='o-grid-item3__lineitems'>
        <div className='c-line-items__colour  u-bold'>
          <span>Colour</span>
        </div>
        <div className='c-line-items__size  u-bold'>
          <span>Option</span>
        </div>
        <div className='c-line-items__quantity u-bold'>
          <span>Quantity</span>
        </div>
      </div>
      <div className='o-grid-item4__lineitems'>
        <span className='c-line-items__space'>Grey</span>
        <span className='c-line-items__space'>{lineItem.variant}</span>
        <span className='c-line-items__space'>{ this.renderOptions(lineItem) }</span>
      </div>
    </>
  }

  renderLineItems () {
    const cartData = this.props.cart.lineItems.map((lineItem, index) =>
      <div className='o-grid-container__lineitems' key={lineItem.slug}>
        <div className='o-grid-item1__lineitems'>
          <Link href={`/slug?slug=${lineItem.slug}`} as={`${lineItem.canonical_path}`} key={lineItem.slug}>
            <Image className='c-line-items__image' src={lineItem.imageUrl} alt={lineItem.title} key={lineItem.slug} aria-label={lineItem.title} />
          </Link>
        </div>
        { this.renderTitle(lineItem) }
        { this.renderParams(lineItem) }
        { this.renderButtonsAndTotal(lineItem) }
      </div>
    )
    return cartData
  }

  render () {
    const cart = this.props.cart

    if (cart.totalQuantity === 0) {
      return null
    } else {
      return <div className='c-line-items'>
        { this.renderLineItems() }
      </div>
    }
  }
}

export default LineItems
