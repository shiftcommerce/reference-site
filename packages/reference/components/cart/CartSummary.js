import { Component } from 'react'

export default class CartSummary extends Component {
  calculateSummary () {
    const lineItems = this.props.cart.lineItems
    let [ total, discount, tax, subTotal ] = [ 0, 0, 0, 0 ]

    lineItems.map((li, index) => {
      subTotal += (li.quantity * li.price)
      discount += (li.quantity * (li.discount || 0))
    })
    total = subTotal - discount + tax
    return { subTotal, total, tax, discount }
  }

  render () {
    const { subTotal, total, tax, discount } = this.calculateSummary()

    return (
      <section className='c-cart-summary'>
        <dl>
          <dt> Sub total: </dt>
          <dd> &pound;{ subTotal } </dd>
        </dl>
        <dl>
          <dt> Tax applied: </dt>
          <dd className='u-text-color--red'> +&pound;{ tax } </dd>
        </dl>
        <dl>
          <dt> Amount you saved on this purchase: </dt>
          <dd className='u-text-color--green'> -&pound;{ discount } </dd>
        </dl>
        <dl>
          <dt> Total amount to be paid: </dt>
          <dd> <b>&pound;{ total }</b> </dd>
        </dl>
      </section>
    )
  }
}
