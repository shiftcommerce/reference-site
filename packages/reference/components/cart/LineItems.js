import { Component } from 'react'
import Link from 'next/link'

// Objects
import Image from '../../objects/Image'

export default class LineItems extends Component {
  renderOptions (lineItem) {
    return <select value={lineItem.quantity} onChange={this.props.updateQuantity} data-variant={lineItem.sku}>
      {
        [ ...Array(Math.max(lineItem.quantity, 11)) ].map((key, index) =>
          <option value={index} key={index} aria-setsize={index} aria-posinset={index + 1}>
            { index }
          </option>
        )
      }
    </select>
  }

  renderLineItems () {
    const cartData = this.props.cart.lineItems.map((lineItem, index) =>
      <tr key={index} className='c-line-items__column'>
        <td aria-label={lineItem.title}>
          <Link href={`/products/${lineItem.productSku}`}>
            <a>
              <Image src={lineItem.imageUrl} alt={lineItem.title} className='c-line-items__image' aria-label={lineItem.title} />
              { lineItem.title }
            </a>
          </Link>
        </td>
        <td>
          { lineItem.size }
        </td>
        <td>
          { this.renderOptions(lineItem) }
        </td>
        <td>
          -&pound;{ (lineItem.discount || 0) * lineItem.quantity }
        </td>
        <td>
          &pound;{ lineItem.price * lineItem.quantity }
        </td>
      </tr>
    )
    return cartData
  }

  render () {
    const cart = this.props.cart

    if (cart.lineItems.length === 0) {
      return null
    } else {
      return <table className='c-line-items'>
        <thead>
          <tr className='c-line-items__header'>
            <th scope='col'>
              Your Items
            </th>
            <th scope='col'>
              Size
            </th>
            <th scope='col'>
              Quantity
            </th>
            <th scope='col'>
              You save
            </th>
            <th scope='col'>
              Subtotal
            </th>
          </tr>
        </thead>
        <tbody>
          { this.renderLineItems() }
        </tbody>
      </table>
    }
  }
}
