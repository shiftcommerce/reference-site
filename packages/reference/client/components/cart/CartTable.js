// Libraries
import { Component } from 'react'
import Pluralize from 'react-pluralize'

// Lib
import { fixedPrice } from '../../lib/fixedPrice'
import { calculateCartSummary } from '../../lib/calculateCartSummary'

// Components
import LineItems from './LineItems'
import CartSummary from './CartSummary'
import CartNoData from './CartNoData'

class CartTable extends Component {
  renderCartData () {
    const cart = this.props.cart

    if (cart.totalQuantity === 0) {
      return <CartNoData />
    } else {
      return <section className='u-float-clear'>
        <LineItems {...this.props} aria-label='Line Items' />
        <CartSummary {...this.props} aria-label='Cart Summary' />
      </section>
    }
  }

  render () {
    const { cart } = this.props

    const totals = calculateCartSummary(cart)

    return (
      <section className='c-cart-table c-cart-table--margin'>
        <section className='c-cart-table__header'>
          <div className='u-float--left'>
            <h1 className='c-cart-table__title'> Your Shopping Basket <a className='c-cart-table__amount'>({ cart.totalQuantity })</a> </h1>
            <p className='c-cart-table__description'>You have <a>{ cart.totalQuantity }</a> <Pluralize singular='item' count={cart.totalQuantity} showCount={false} /> in your shopping basket</p>
            <p className='c-cart-table__description'><Pluralize singular='This' plural='These' count={cart.totalQuantity} showCount={false} /> <Pluralize singular='item' count={cart.totalQuantity} showCount={false} /> will be saved for 48 hours depending on availablility</p>
          </div>
          <div>
            <h4 className='c-cart-table__total'>&pound;{ fixedPrice(totals.total) }</h4>
          </div>
        </section>
        { this.renderCartData() }
      </section>
    )
  }
}

export default CartTable
