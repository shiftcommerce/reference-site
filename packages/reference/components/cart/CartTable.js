import { Component } from 'react'

// Components
import LineItems from './LineItems'
import CartSummary from './CartSummary'
import CartNoData from './CartNoData'

export default class CartTable extends Component {
  renderCartData () {
    const cart = this.props.cart

    if (cart.lineItems.length === 0) {
      return <CartNoData />
    } else {
      return <section className='u-float-clear'>
        <LineItems { ...this.props } aria-label='Line Items' />
        <CartSummary { ...this.props } aria-label='Cart Summary' />
      </section>
    }
  }
  render () {
    const {
      cart
    } = this.props
    const itemsKey = cart.lineItems.length === 1 ? 'item' : 'items'

    return (
      <section>
        <section className='c-cart-table__header'>
          <div className='u-float--left'>
            <h1> My Bag </h1>
          </div>
          <div className='u-float--right'>
            <h4> <b>{ cart.lineItems.length } { itemsKey }</b> in your bag </h4>
          </div>
        </section>
        { this.renderCartData() }
      </section>
    )
  }
}
