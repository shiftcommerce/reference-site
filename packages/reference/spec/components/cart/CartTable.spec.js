import CartTable from '../../../components/cart/CartTable'
import CartNoData from '../../../components/cart/CartNoData'
import LineItems from '../../../components/cart/LineItems'
import CartSummary from '../../../components/cart/CartSummary'

test('renders the exception messages when no line items are available', () => {
  // arrange
  const cart = {
    lineItems: []
  }

  const updateQuantity = () => {}

  // act
  const wrapper = mount(
    <CartTable cart={cart} />
  )

  // assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('My Bag')
  expect(wrapper).toIncludeText('0 items in your bag')
  expect(wrapper).toContainReact(<CartNoData />)
  expect(wrapper).not.toContainReact(<LineItems cart={cart} updateQuantity={updateQuantity} aria-label='Line Items' />)
  expect(wrapper).not.toContainReact(<CartSummary cart={cart} updateQuantity={updateQuantity} aria-label='Cart Summary' />)
})

test('renders the correct message when line items are available in cart', () => {
  // arrange
  const cart = {
    lineItems: [
      {
        title: 'test',
        price: 10,
        discount: 0,
        quantity: 2,
        sku: '123',
        imageUrl: '',
        size: 'size - 8'
      }
    ]
  }

  const updateQuantity = () => {}

  // act
  const wrapper = mount(
    <CartTable cart={cart} updateQuantity={updateQuantity} />
  )

  // assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('My Bag')
  expect(wrapper).toIncludeText('1 item in your bag')
  expect(wrapper).not.toContainReact(<CartNoData />)
  expect(wrapper).toContainReact(<LineItems cart={cart} updateQuantity={updateQuantity} aria-label='Line Items' />)
  expect(wrapper).toContainReact(<CartSummary cart={cart} updateQuantity={updateQuantity} aria-label='Cart Summary' />)
})
