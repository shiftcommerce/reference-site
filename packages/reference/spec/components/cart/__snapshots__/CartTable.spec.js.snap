// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct message when line items are available in cart 1`] = `
<CartTable
  cart={
    Object {
      "lineItems": Array [
        Object {
          "discount": 0,
          "imageUrl": "",
          "price": 10,
          "quantity": 2,
          "size": "size - 8",
          "sku": "123",
          "title": "test",
        },
      ],
    }
  }
  updateQuantity={[Function]}
>
  <section>
    <section
      className="c-cart-table__header"
    >
      <div
        className="u-float--left"
      >
        <h1>
           My Bag 
        </h1>
      </div>
      <div
        className="u-float--right"
      >
        <h4>
           
          <b>
            1
             
            item
          </b>
           in your bag 
        </h4>
      </div>
    </section>
    <section
      className="u-float-clear"
    >
      <LineItems
        aria-label="Line Items"
        cart={
          Object {
            "lineItems": Array [
              Object {
                "discount": 0,
                "imageUrl": "",
                "price": 10,
                "quantity": 2,
                "size": "size - 8",
                "sku": "123",
                "title": "test",
              },
            ],
          }
        }
        updateQuantity={[Function]}
      >
        <table
          className="c-line-items"
        >
          <thead>
            <tr
              className="c-line-items__header"
            >
              <th
                scope="col"
              >
                Your Items
              </th>
              <th
                scope="col"
              >
                Size
              </th>
              <th
                scope="col"
              >
                Quantity
              </th>
              <th
                scope="col"
              >
                You save
              </th>
              <th
                scope="col"
              >
                Subtotal
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              className="c-line-items__column"
            >
              <td
                aria-label="test"
              >
                <Link
                  href="/products/undefined"
                >
                  <a
                    href="/products/undefined"
                    onClick={[Function]}
                  >
                    <Image
                      alt="test"
                      aria-label="test"
                      className="c-line-items__image"
                      src=""
                    >
                      <div
                        className="c-dummy_image c-line-items__image"
                        style={
                          Object {
                            "backgroundSize": 150,
                            "height": undefined,
                            "width": undefined,
                          }
                        }
                      />
                    </Image>
                    test
                  </a>
                </Link>
              </td>
              <td>
                size - 8
              </td>
              <td>
                <select
                  data-variant="123"
                  onChange={[Function]}
                  value={2}
                >
                  <option
                    aria-posinset={1}
                    aria-setsize={0}
                    value={0}
                  >
                    0
                  </option>
                  <option
                    aria-posinset={2}
                    aria-setsize={1}
                    value={1}
                  >
                    1
                  </option>
                  <option
                    aria-posinset={3}
                    aria-setsize={2}
                    value={2}
                  >
                    2
                  </option>
                  <option
                    aria-posinset={4}
                    aria-setsize={3}
                    value={3}
                  >
                    3
                  </option>
                  <option
                    aria-posinset={5}
                    aria-setsize={4}
                    value={4}
                  >
                    4
                  </option>
                  <option
                    aria-posinset={6}
                    aria-setsize={5}
                    value={5}
                  >
                    5
                  </option>
                  <option
                    aria-posinset={7}
                    aria-setsize={6}
                    value={6}
                  >
                    6
                  </option>
                  <option
                    aria-posinset={8}
                    aria-setsize={7}
                    value={7}
                  >
                    7
                  </option>
                  <option
                    aria-posinset={9}
                    aria-setsize={8}
                    value={8}
                  >
                    8
                  </option>
                  <option
                    aria-posinset={10}
                    aria-setsize={9}
                    value={9}
                  >
                    9
                  </option>
                  <option
                    aria-posinset={11}
                    aria-setsize={10}
                    value={10}
                  >
                    10
                  </option>
                </select>
              </td>
              <td>
                -£
                0
              </td>
              <td>
                £
                20
              </td>
            </tr>
          </tbody>
        </table>
      </LineItems>
      <CartSummary
        aria-label="Cart Summary"
        cart={
          Object {
            "lineItems": Array [
              Object {
                "discount": 0,
                "imageUrl": "",
                "price": 10,
                "quantity": 2,
                "size": "size - 8",
                "sku": "123",
                "title": "test",
              },
            ],
          }
        }
        updateQuantity={[Function]}
      >
        <section
          className="c-cart-summary"
        >
          <dl>
            <dt>
               Sub total: 
            </dt>
            <dd>
               £
              20
               
            </dd>
          </dl>
          <dl>
            <dt>
               Tax applied: 
            </dt>
            <dd
              className="u-text-color--red"
            >
               +£
              0
               
            </dd>
          </dl>
          <dl>
            <dt>
               Amount you saved on this purchase: 
            </dt>
            <dd
              className="u-text-color--green"
            >
               -£
              0
               
            </dd>
          </dl>
          <dl>
            <dt>
               Total amount to be paid: 
            </dt>
            <dd>
               
              <b>
                £
                20
              </b>
               
            </dd>
          </dl>
        </section>
      </CartSummary>
    </section>
  </section>
</CartTable>
`;

exports[`renders the exception messages when no line items are available 1`] = `
<CartTable
  cart={
    Object {
      "lineItems": Array [],
    }
  }
>
  <section>
    <section
      className="c-cart-table__header"
    >
      <div
        className="u-float--left"
      >
        <h1>
           My Bag 
        </h1>
      </div>
      <div
        className="u-float--right"
      >
        <h4>
           
          <b>
            0
             
            items
          </b>
           in your bag 
        </h4>
      </div>
    </section>
    <CartNoData>
      <section
        className="c-cart-no-data"
      >
        <p>
          You have no items in your cart. If you are expecting items in it, please refresh.
        </p>
        <p>
          If issue still persists, please check with our customer care.
        </p>
      </section>
    </CartNoData>
  </section>
</CartTable>
`;
